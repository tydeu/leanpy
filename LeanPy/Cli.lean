/-
Copyright (c) 2025 Mac Malone. All rights reserved.
Released under the Apache 2.0 license as described in the file LICENSE.
syntax Authors Mac Malone
-/
import LeanPy.Cli.EvalFile

open Lake

namespace LeanPy

def cli  (args : List String) : IO UInt32 := do
  let [file] := args
    | IO.eprintln "error: missing file name"
      return 1
  let messages ← evalPyFile file
  messages.unreported.forM fun msg => do
    IO.eprint (← mkMessageString msg)
  return if messages.hasErrors then 1 else 0
