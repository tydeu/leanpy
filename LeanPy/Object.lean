/-
Copyright (c) 2025 Mac Malone. All rights reserved.
Released under the Apache 2.0 license as described in the file LICENSE.
syntax Authors Mac Malone
-/
import Std.Data.HashMap

namespace LeanPy

/-- The name of an object's Lean data type. -/
abbrev DataTy := Lean.Name

private opaque ObjectData.nonemptyType (ty : DataTy) : NonemptyType.{0}

/-- Per-type data. -/
def ObjectData (ty : DataTy) : Type := (ObjectData.nonemptyType ty).type

instance ObjectData.instNonempty : Nonempty (ObjectData ty) :=
  (ObjectData.nonemptyType ty).property

open TypeName

set_option linter.unusedVariables.funArgs false in
def ObjectData.mk [TypeName α] (a : α) (h : typeName α = n) : ObjectData n :=
  unsafe unsafeCast a

unsafe def ObjectData.getImpl [Nonempty α] [TypeName α] (self : ObjectData (typeName α)) : α :=
  unsafeCast self

@[implemented_by getImpl]
opaque ObjectData.get [Nonempty α] [TypeName α] (self : ObjectData (typeName α)) : α

/-- A Python type object. -/
-- TODO: Enrich with proper fields
structure TypeObject where
  name : String
  dataTy : DataTy
  deriving Nonempty

/-- A Python type object with a known Lean representation. -/
structure DTypeObject (α : Type u) [TypeName α] extends TypeObject where
  dataTy := typeName α
  dataTy_eq_typeName : dataTy = typeName α := by exact rfl

instance [TypeName α] : CoeOut (DTypeObject α) TypeObject :=
  ⟨DTypeObject.toTypeObject⟩

/-- A Python object. -/
structure Object where
  mk' ::
    ty : TypeObject
    data : ObjectData ty.dataTy

def Object.mk [TypeName α] (ty : DTypeObject α) (data : α) : Object :=
  ⟨ty, ObjectData.mk data ty.dataTy_eq_typeName.symm⟩

@[inline] def Object.dataTy (self : Object) : DataTy :=
  self.ty.dataTy

structure DObject (α : Type u) [TypeName α] extends Object where
  toObject_dataTy_eq_typeName : toObject.dataTy = typeName α

instance [TypeName α] : CoeOut (DObject α) Object :=
  ⟨DObject.toObject⟩

deriving instance TypeName for Unit

def noneType : DTypeObject Unit where
  name := "NoneType"

def none : Object :=
  .mk noneType ()

def Object.isNone (self : Object) : Bool :=
  self.ty.name == noneType.name

deriving instance TypeName for Int

def boolType : DTypeObject Int where
  name := "bool"

protected def Object.false : Object :=
  .mk boolType 0

instance : CoeDep Bool false Object := ⟨Object.false⟩

protected def Object.true : Object :=
  .mk boolType 1

instance : CoeDep Bool true Object := ⟨Object.true⟩
