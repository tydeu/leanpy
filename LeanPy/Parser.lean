/-
Copyright (c) 2023 Mac Malone. All rights reserved.
Released under the Apache 2.0 license as described in the file LICENSE.
syntax Authors Mac Malone
-/
import Lean.Parser.Basic
import Lean.Parser.Syntax

/-! # Parser & Syntax Utilities
Additional definitions to help with defining the Python grammar.
-/

namespace LeanPy
open Lean Parser Command PrettyPrinter Formatter Parenthesizer

/--
`$`-style application grouping for the syntax category.

For example, `withPostion $ x` expands to `withPosition(x)`.
-/
macro id:ident " $ " xs:stx*: stx => `(stx| $id:ident($[$xs]*))

/--
A convenient multiline syntax for defining large alternatives. For example,

```
syntax foobar
  | foo
  | bar
  | baz
```

is equivalent to

```
syntax foobar := foo <|> (bar <|> baz)
```
-/
syntax (name := syntaxAbbrevAlts) (docComment)?
"syntax" ident ppIndent(many1Indent(ppLine "| " ppSpace stx+)) : command

macro_rules
| `(syntaxAbbrevAlts| $(doc?)? syntax $id $[| $[$alts:stx]*]*) => do
  let alt0 := ⟨mkNullNode alts.back⟩
  let stx ← alts[0:alts.size-1].foldrM (init := alt0) fun xs x =>
    `(stx|($[$xs]*) <|> $x)
  `(syntaxAbbrev| $[$doc?:docComment]? syntax $id := $stx)

/-- A `Parser` for `rawFn`. -/
def raw (fn : ParserFn) (trailingWs := false) : Parser where
  fn := rawFn fn trailingWs

@[combinator_formatter raw] def raw.formatter :=
  Formatter.visitAtom Name.anonymous

@[combinator_parenthesizer raw] def raw.parenthesizer :=
  Parenthesizer.visitToken

/--
Check if the topmost atom of the syntax stack matches `f`. If not, error with `msg`.
If the syntax on top of the stack is a node, look at its final argument recursively.
-/
@[run_parser_attribute_hooks] def checkAtomTop (f : String → Bool) (msg : String) :=
  checkStackTop go msg
where
  go
  | .missing => false
  | .ident .. => false
  | .atom (val := val) .. => f val
  | .node (args := args) .. =>
      if h : args.size > 0 then
        go <| args[args.size - 1]'(Nat.sub_succ_lt_self _ _ h)
      else
        false

@[run_parser_attribute_hooks]
def commaBefore := checkAtomTop (· = ",") "expected ',' before"

def simpleIdentFn : ParserFn := fun ctx st =>
  let startPos := st.pos
  let initStackSz := st.stackSize
  let expected := ["expected simple unqualified identifier"]
  let st := tokenFn expected ctx st
  if st.hasError then
    st
  else
    match st.stxStack.back with
    | Syntax.ident _ _ (.str .anonymous _) _ => st
    | _  => st.mkErrorsAt expected startPos initStackSz

def simpleIdentNoAntiquot : Parser where
  fn   := simpleIdentFn
  info := mkAtomicInfo "ident"

@[combinator_formatter simpleIdentNoAntiquot]
def simpleIdentNoAntiquot.formatter := identNoAntiquot.formatter

@[combinator_parenthesizer simpleIdentNoAntiquot]
def simpleIdentNoAntiquot.parenthesizer := identNoAntiquot.parenthesizer

/--
Parses an identifier that is a simple name (i.e., `Name.mkSimple`).
That is, it parses an identifier without a `.` (an unqualified/atomic identifier).
-/
@[run_parser_attribute_hooks] def simpleIdent : Parser :=
  withAntiquot (mkAntiquot "ident" identKind) simpleIdentNoAntiquot
