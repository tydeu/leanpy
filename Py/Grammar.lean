/-
Copyright (c) 2023 Mac Malone. All rights reserved.
Released under the Apache 2.0 license as described in the file LICENSE.
syntax Authors Mac Malone
-/
import LeanPy.Parser

/-!
# The Python Grammar

This module contains a Lean DSL that encodes the Python grammar.
It uses Python 3's [Full Grammar Specification][1] as a guideline.

[1]: https://docs.python.org/3/reference/grammar.html
-/

namespace Py.Grammar
open LeanPy

/-- Use `[stx]` as an alternative to `(stx)?` following the specification style. -/
local macro "[" xs:stx* "]" : stx => `(stx|($[$xs:stx]*)?)

/-! ## Basic Tokens -/

syntax name := simpleIdent
syntax dottedName := ident -- name ("." name)*
syntax attr := ident -- name ("." name)+
syntax nameOrAttr := ident -- name ("." name)*
syntax lambdaParam := name

/-! ## Expressions -/

declare_syntax_cat pyExpr (behavior := symbol)

syntax typeComment := "# " "type: " pyExpr

syntax exprs := pyExpr,+,?
syntax starredExpr := "*" pyExpr
syntax doubleStarredExpr := "**" pyExpr

syntax starExpr
  | "*" pyExpr:55
  | pyExpr

syntax starExprs := starExpr,+,?

syntax yieldExpr := "yield" (("from" pyExpr) <|> starExpr,*,?)

syntax assignmentExpr := atomic(name " := ") pyExpr

syntax namedExpr
  | assignmentExpr
  | pyExpr !":="

syntax starNamedExpr
  | "*" pyExpr:55
  | namedExpr

/-! ### Logical Expressions -/

syntax (name := condExpr) pyExpr:30 colGt " if " pyExpr:30 colGt " else " pyExpr : pyExpr
syntax:30 (name := disjunction) pyExpr:31 (" or " pyExpr:31)+ : pyExpr
syntax:35 (name := conjunction) pyExpr:36 (" and " pyExpr:36)+ : pyExpr
syntax:38 (name := inversion) "not " pyExpr:40 : pyExpr

/-! ### Comparison Operators -/

syntax:50 (name := eq)    pyExpr:50 " == " pyExpr:51 : pyExpr
syntax:50 (name := ne)    pyExpr:50 " != " pyExpr:51 : pyExpr
syntax:50 (name := le)    pyExpr:50 " <= " pyExpr:51 : pyExpr
syntax:50 (name := lt)    pyExpr:50 " < " pyExpr:51 : pyExpr
syntax:50 (name := ge)    pyExpr:50 " >= " pyExpr:51 : pyExpr
syntax:50 (name := gt)    pyExpr:50 " > " pyExpr:51 : pyExpr
syntax:50 (name := notIn) pyExpr:50 " not " "in " pyExpr:51 : pyExpr
syntax:50 (name := «in»)  pyExpr:50 " in " pyExpr:51 : pyExpr
syntax:50 (name := isNot) pyExpr:50 " is " "not " pyExpr:51 : pyExpr
syntax:50 (name := «is»)  pyExpr:50 " is " pyExpr:51 : pyExpr

/-! ### Bitwise Operators -/

syntax:55 (name := bitwiseOr)  pyExpr:55 " | " pyExpr:56 : pyExpr
syntax:58 (name := bitwiseXor) pyExpr:58 " ^ " pyExpr:59 : pyExpr
syntax:60 (name := bitwiseAnd) pyExpr:60 " & " pyExpr:61 : pyExpr

/- shift_expr -/
syntax:63 (name := shl) pyExpr:63 "<<" pyExpr:65 : pyExpr
syntax:63 (name := shr) pyExpr:63 ">>" pyExpr:65 : pyExpr

/-! ### Arithmetic Operators -/

/- sum -/
syntax:65 (name := add) pyExpr:65 " + " pyExpr:66 : pyExpr
syntax:65 (name := sub) pyExpr:65 " - " pyExpr:66 : pyExpr

/- term -/
syntax:70 (name := mul)     pyExpr:70 " * " pyExpr:71 : pyExpr
syntax:70 (name := tdiv)    pyExpr:70 " / " pyExpr:71 : pyExpr
syntax:70 (name := fdiv)    pyExpr:70 " // " pyExpr:71 : pyExpr
syntax:70 (name := mod)     pyExpr:70 " % " pyExpr:71 : pyExpr
syntax:70 (name := matmul)  pyExpr:70 " @ " pyExpr:71 : pyExpr

/- factor -/
syntax:75 (name := pos) "+" pyExpr:75 : pyExpr
syntax:75 (name := neg) "-" pyExpr:75 : pyExpr
syntax:75 (name := cmp) "~" pyExpr:75 : pyExpr

/- power -/
syntax:75 (name := power) pyExpr:76 "**" pyExpr:75 : pyExpr

/-! ### Primary Elements -/

syntax:100 (name := await) "await " pyExpr:101 : pyExpr

syntax slice
  | assignmentExpr
  | [pyExpr] ":" [pyExpr] [":" [pyExpr]]
  | pyExpr !":="

syntax slices
  | atomic(slice !",")
  | (slice <|> starredExpr),+,?

syntax:arg pyExpr:arg "[" slices "]" : pyExpr

/- Captures both simple names and member accesses of names. -/
syntax:max (name := identExpr) ident : pyExpr

syntax:arg pyExpr:arg "." name : pyExpr

syntax:max (name := trueExpr)   "True" : pyExpr
syntax:max (name := falseExpr)  "False" : pyExpr
syntax:max (name := noneExpr)   "None" : pyExpr
syntax:max (name := numExpr)    num : pyExpr
syntax:max (name := group)      "(" (yieldExpr <|> namedExpr) ")" : pyExpr
syntax:max (name := ellipsis)   "..." : pyExpr

/-! ### Lambda Functions -/

/-
NOTE FROM THE SPECIFICATION:
`lambdaParams` etc. duplicates parameters but without annotations
or type comments, and if there's no comma after a parameter, we expect
a colon, not a close parenthesis.  (For more, see parameters above.)
-/

syntax default := " = " pyExpr

syntax lambdaParamNoDefault :=
  lambdaParam ("," <|> lookahead(":"))

syntax lambdaParamWithDefault :=
  lambdaParam default ("," <|> lookahead(":"))

syntax lambdaParamMaybeDefault :=
  lambdaParam [default] ("," <|> lookahead(":"))

syntax lambdaKwds :=
  "**" lambdaParamNoDefault

syntax lambdaStarEtc
  | atomic("*" ",") lambdaParamMaybeDefault+ [lambdaKwds]
  | "*" lambdaParamNoDefault lambdaParamMaybeDefault* [lambdaKwds]
  | lambdaKwds

syntax lambdaSlashNoDefault
  | lambdaParamNoDefault+ "/" ("," <|> lookahead(":"))

syntax lambdaSlashWithDefault
  | lambdaParamNoDefault* lambdaParamWithDefault+ "/" ("," <|> lookahead(":"))

syntax lambdaParams
  | lambdaSlashNoDefault lambdaParamNoDefault* lambdaParamWithDefault* [lambdaStarEtc]
  | lambdaSlashWithDefault lambdaParamWithDefault* [lambdaStarEtc]
  | lambdaParamNoDefault+ lambdaParamWithDefault* [lambdaStarEtc]
  | lambdaParamWithDefault+ [lambdaStarEtc]
  | lambdaStarEtc

syntax:20 (name := lambdef) "lambda " [lambdaParams] ":" pyExpr : pyExpr

/- ## Literals -/

syntax:max (name := strings) str+ : pyExpr
syntax:max (name := list)   "[" starNamedExpr,*,? "]" : pyExpr
syntax:max (name := tuple)  "(" [starNamedExpr "," starNamedExpr,*,?] ")" : pyExpr
syntax:max (name := set)    "{" starNamedExpr,+,? "}" : pyExpr

/- ### Dicts -/

syntax kvpair := pyExpr ":" pyExpr
syntax doubleStarredKvpair := (("**" pyExpr:55) <|> kvpair)
syntax:max (name := dict) "{" doubleStarredKvpair,*,? "}" : pyExpr

/- ### Comprehensions & Generators -/

declare_syntax_cat pyStarTarget
syntax starTargets := pyStarTarget,+,?

syntax forIfClause :=
  atomic("async "? "for " starTargets " in ") disjunction (" if " disjunction )*

syntax forIfClauses := forIfClause+

syntax:max (name := listcomp) "[" namedExpr forIfClauses "]" : pyExpr
syntax:max (name := setcomp)  "{" namedExpr forIfClauses "}" : pyExpr
syntax:max (name := dictcomp) "{" kvpair forIfClauses "}" : pyExpr

syntax:max (name := genexp)   "(" (assignmentExpr <|> (pyExpr !":=")) forIfClauses ")" : pyExpr
syntax:arg pyExpr:arg genexp : pyExpr

/- ### Function Calls -/

syntax kwarg := atomic(name " = ") pyExpr

syntax kwargs
  | lookahead("**") (doubleStarredExpr <|> kwarg),+,?
  | (starredExpr <|> kwarg),+,? [commaBefore (doubleStarredExpr <|> kwarg),+,?]

syntax args
  | (!("**" <|> kwarg) (starredExpr <|> assignmentExpr <|> (pyExpr !":=")) !"="),+,?
    [commaBefore kwargs]
  | kwargs

syntax arguments := args lookahead(")")
syntax:arg (name := callExpr) pyExpr:arg "(" [arguments] ")" : pyExpr

/-! ## Assignment Targets -/

/-
NOTE FROM THE SPECIFICATION:
Star targets may contain `*bitwiseOr`, targets may not.
-/

syntax tLookahead := "(" <|> "[" <|> "."

declare_syntax_cat pyTPrimary
syntax pyTPrimary "." name lookahead(tLookahead) : pyTPrimary
syntax pyTPrimary "[" slices "]" lookahead(tLookahead) : pyTPrimary
syntax pyTPrimary genexp lookahead(tLookahead) : pyTPrimary
syntax pyTPrimary "(" [arguments] ")" lookahead(tLookahead) : pyTPrimary
syntax pyExpr:max lookahead(tLookahead) : pyTPrimary

syntax singleSubscriptAttributeTarget :=
  pyTPrimary (("." name) <|> ("[" slices "]")) !tLookahead

syntax "*" !"*" pyStarTarget : pyStarTarget
syntax:arg singleSubscriptAttributeTarget : pyStarTarget
syntax:max "(" pyStarTarget,*,? ")" : pyStarTarget
syntax:max "[" pyStarTarget,*,? "]" : pyStarTarget
syntax:max name : pyStarTarget

declare_syntax_cat pySingleTarget
syntax "(" pySingleTarget ")" : pySingleTarget
syntax singleSubscriptAttributeTarget : pySingleTarget
syntax name : pySingleTarget

/-! ### Targets for `del` Statements -/

declare_syntax_cat pyDelTarget
syntax delTargets := pyDelTarget,+,?
syntax singleSubscriptAttributeTarget : pyDelTarget

/- del_t_atom -/
syntax:max name : pyDelTarget
syntax:max "(" pyDelTarget,*,? ")" : pyDelTarget
syntax:max "[" pyDelTarget,*,? "]" : pyDelTarget

/-! ## Typing Elements -/

-- NOTE FROM THE SPECIFICATION:  `typeExprs` allow `*`/`**` but ignore them
syntax typeExprs
  | "**" pyExpr
  | "*" pyExpr ["," "**" pyExpr]
  | pyExpr,+,? [commaBefore "*" pyExpr ("," <|> !"**")] [commaBefore "**" pyExpr]

syntax funcTypeComment
  | linebreak typeComment colGt -- SPEC NOTE: Must be followed by indented block
  | typeComment

/-! ## Statements -/

declare_syntax_cat pySimpleStmt (behavior := symbol)
syntax simpleStmts := withPosition(sepBy1(pySimpleStmt, "; ")) linebreak

declare_syntax_cat pyCompoundStmt (behavior := symbol)
syntax stmt := pyCompoundStmt <|> simpleStmts

syntax stmtNewline
  | pyCompoundStmt linebreak
  | simpleStmts
  | linebreak

/-! ### Simple Statements -/

syntax returnStmt   := "return " starExpr,*,?
syntax raiseStmt    := "raise " [pyExpr ["from" pyExpr]]
syntax passStmt     := "pass"
syntax delStmt      := "del " delTargets lookahead(";" <|> linebreak)
syntax yieldStmt    := yieldExpr
syntax assertStmt   := "assert " pyExpr ["," pyExpr]
syntax breakStmt    := "break"
syntax continueStmt := "continue"
syntax globalStmt   := "global " name,+
syntax nonlocalStmt := "nonlocal " name,+

attribute [pySimpleStmt_parser] returnStmt
attribute [pySimpleStmt_parser] raiseStmt
attribute [pySimpleStmt_parser] passStmt
attribute [pySimpleStmt_parser] delStmt
attribute [pySimpleStmt_parser] yieldStmt
attribute [pySimpleStmt_parser] assertStmt
attribute [pySimpleStmt_parser] breakStmt
attribute [pySimpleStmt_parser] continueStmt
attribute [pySimpleStmt_parser] globalStmt
attribute [pySimpleStmt_parser] nonlocalStmt

/-! ### Assignment Statements -/

syntax augassign
  | "+="
  | "-="
  | "*="
  | "@="
  | "/="
  | "%="
  | "&="
  | "|="
  | "^="
  | "<<="
  | ">>="
  | "**="
  | "//="

syntax annotatedRhs := yieldExpr <|> starExprs

/-
NOTE FROM THE SPECIFICATION:
`annotatedRhs` may start with `yield`; `yieldExpr` must start with `yield`
-/
syntax assignment
  | atomic(name ": ") pyExpr [" = " annotatedRhs]
  | atomic(("(" pySingleTarget ")" <|> singleSubscriptAttributeTarget) ": ")
      pyExpr ["=" annotatedRhs]
  | (atomic(starTargets) "=")+ annotatedRhs !"=" [typeComment]
  | atomic(pySingleTarget augassign) annotatedRhs

/-
NOTE: Assignment MUST be higher priority than expression in the grammar,
otherwise parsing a simple assignment will break (and this is a known feature
of the Python grammar).
-/
attribute [pySimpleStmt_parser] starExprs
attribute [pySimpleStmt_parser high] assignment

/-! ### Import Statements -/

syntax importFromAsName := name ["as" name]
syntax dottedAsName := dottedName ["as" name]
syntax importName := "import " dottedAsName,+

syntax importFromTargets
  | "(" importFromAsName,+,? ")"
  | importFromAsName,+
  | "*"

@[run_parser_attribute_hooks]
def dots := raw <| Lean.Parser.takeWhileFn (· = '.')

syntax importFrom :=
  "from " dots dottedName " import " importFromTargets

attribute [pySimpleStmt_parser] importName
attribute [pySimpleStmt_parser] importFrom

/-! ### Compound Statements -/

syntax block
  | ppSpace ("\\" <|> lineEq) simpleStmts
  | ppIndent(colGt withPosition(many1Indent(ppLine stmt)))

syntax decorator := "@" namedExpr linebreak

/-! ### Class Definitions -/

syntax classDefRaw :=
  "class " name ["(" [arguments] ")"] ":" block

syntax classDef :=
  withPosition(decorator* classDefRaw)

attribute [pyCompoundStmt_parser] classDef

/-! ### Function Definitions -/

/-
NOTE FROM THE SPECIFICATION:

One parameter.  This *includes* a following comma and type comment.

There are three styles:
- No default
- With default
- Maybe with default

There are two alternative forms of each, to deal with type comments:
- Ends in a comma followed by an optional type comment
- No comma, optional type comment, must be followed by close paren
The latter form is for a final parameter without trailing comma.
-/

syntax annotation := ":" pyExpr
syntax param := atomic(name [annotation])
syntax starAnnotation := ":" starExpr
syntax paramStarAnnotation := atomic(name starAnnotation)

syntax paramSep :=
  ("," [typeComment]) <|> ([typeComment] lookahead(")"))

syntax paramNoDefault := param paramSep
syntax paramNoDefaultStarAnnotation := paramStarAnnotation paramSep
syntax paramWithDefault := atomic(param default) paramSep
syntax paramMaybeDefault := param [default] paramSep

syntax slashNoDefault :=
  paramNoDefault+ "/" ("," <|> lookahead(")"))

syntax slashWithDefault :=
  paramNoDefault* paramWithDefault+ "/" ("," <|> lookahead(")"))

syntax kwds :=
  "**" paramNoDefault

syntax starEtc
  | "*" paramNoDefault paramMaybeDefault* [kwds]
  | "*" paramNoDefaultStarAnnotation paramMaybeDefault* [kwds]
  | "*" "," paramMaybeDefault+ [kwds]
  | kwds

syntax params
  | slashNoDefault paramNoDefault* paramWithDefault* [starEtc]
  | slashWithDefault paramWithDefault* [starEtc]
  | paramNoDefault+ paramWithDefault* [starEtc]
  | paramWithDefault+ [starEtc]
  | starEtc

syntax functionDefRaw :=
  "async "? "def " name "(" [params] ")" [" -> " pyExpr] ":" [funcTypeComment] block

syntax functionDef :=
  withPosition(decorator* functionDefRaw)

attribute [pyCompoundStmt_parser] functionDef

/-! ### If Statement -/

syntax elseBlock := "else" ":" block
syntax elifStmt := "elif " namedExpr ":" block

syntax ifStmt := withPosition $
  "if " namedExpr ":" block (colEq elifStmt)* [colEq elseBlock]

attribute [pyCompoundStmt_parser] ifStmt

/-! ### While Statement -/

syntax whileStmt := withPosition $
  "while" namedExpr ":" block [colEq elseBlock]

attribute [pyCompoundStmt_parser] whileStmt

/-! ### For Statement -/

syntax forStmt := withPosition $
  atomic("async "? "for" starTargets "in")
  starExprs ":" [typeComment] block [colEq elseBlock]

attribute [pyCompoundStmt_parser] forStmt

/-! ### With Statement -/

syntax withItem :=
  pyExpr ["as" pyStarTarget lookahead("," <|> ")" <|> ":")]

syntax withSig
  | "(" withItem,+,? ")" ":" block
  | withItem,+ ":" [typeComment] block

syntax withStmt :=
  withPosition("async "? "with " withSig)

attribute [pyCompoundStmt_parser] withStmt

/-! ### Try Statement -/

syntax exceptBlock :=
  "except" (":" <|> (pyExpr ["as" name] ":")) block

syntax exceptStarBlock :=
  atomic("except" "*") pyExpr ["as" name] ":" block

syntax finallyBlock :=
  "finally" ":" block

syntax tryStmt := withPosition $
  "try" ":" block colEq (exceptStarBlock+ <|> exceptBlock+)
    [colEq elseBlock] [colEq finallyBlock]

attribute [pyCompoundStmt_parser] tryStmt

/-! ### Match Statement -/

syntax patternCaptureTarget :=
  name !("." <|> "(" <|> "=")

declare_syntax_cat pyPattern (behavior := symbol)

syntax:20 (name := asPattern) pyPattern:21 "as" patternCaptureTarget : pyPattern
syntax:30 (name := orPattern) sepBy1(pyPattern:max,"|") : pyPattern

syntax literalNum :=
  "-"? num [("+" <|> "-") num]

-- Literal exprs are used to restrict permitted mapping pattern keys
syntax literalExpr
  | literalNum
  | strings
  | "None"
  | "True"
  | "False"

syntax:max (name := literalPattern) literalExpr : pyPattern
syntax:max (name := capturePattern) patternCaptureTarget : pyPattern
syntax:max (name := wildcardPattern) "_" : pyPattern
syntax:max (name := valuePattern) attr !("." <|> "(" <|> "=") : pyPattern
syntax:max (name := groupPattern) "(" pyPattern ")" : pyPattern

syntax starPattern :=
  "*" (wildcardPattern <|> patternCaptureTarget)

syntax maybeStarPattern := starPattern <|> pyPattern

syntax sequencePattern
  | "[" maybeStarPattern,*,? "]"
  | "(" [maybeStarPattern "," maybeStarPattern,*,?] ")"

syntax keyValuePattern :=
  (literalExpr <|> attr) ":" pyPattern

syntax itemsPattern := keyValuePattern,+

syntax doubleStarPattern :=
  "**" patternCaptureTarget

syntax mappingPatternParams
  | doubleStarPattern ","?
  | keyValuePattern,+,?  [commaBefore doubleStarPattern ","?]

syntax:max (name := mappingPattern) "{" [mappingPatternParams] "}" : pyPattern

syntax keywordPattern := atomic(name "=") pyPattern

syntax classPatternParams
  |  keywordPattern,+,?
  |  pyPattern,+,? [commaBefore keywordPattern,+,?]

syntax:max (name := classPattern) nameOrAttr "(" [classPatternParams] ")" : pyPattern

syntax patterns
  | starPattern "," maybeStarPattern,*,?
  | pyPattern ["," maybeStarPattern,*,?]

syntax guard := " if " namedExpr

syntax caseBlock := withPosition $
  "case " patterns [guard] ":" block

syntax subjectExpr
  | "*" pyExpr:55 "," starNamedExpr,*,?
  | namedExpr

syntax matchBlock :=
  ppIndent(colGt withPosition(many1Indent(ppLine stmt)))

syntax matchStmt := withPosition $
  "match " subjectExpr ":" matchBlock

attribute [pyCompoundStmt_parser] matchStmt
