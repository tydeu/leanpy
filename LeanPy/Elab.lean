/-
Copyright (c) 2024 Mac Malone. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Mac Malone
-/
import LeanPy.Object
import LeanPy.Grammar
import LeanPy.ElabUtil

open Lean Elab Command Parser

namespace LeanPy

abbrev PyElabM := CoreM
abbrev PyEvalM := PyElabM
abbrev PyEval := Syntax → PyEvalM Object

initialize pyEvalAttribute : KeyedDeclsAttribute PyEval ←
  unsafe mkElabAttribute _ `builtin_py_eval `py_eval `LeanPy.Grammar ``PyEval "py"

def evalPy (stx : Syntax) : PyEvalM Object :=
  elabSyntaxWith stx fun stx => do
    let k := stx.getKind
    let elabFns := pyEvalAttribute.getEntries (← getEnv) k
    if let some obj ← elabSyntaxUsing? stx elabFns then
      return obj
    else
      withInfoTreeContext (mkInfoTree := mkCommandElabInfoTree `no_elab stx) do
        throwError m!"no python evaluator implemented for `{mkConst k}`"

def runPy (stx : Syntax) : PyElabM Unit :=
  withLogging <| discard <| evalPy stx
