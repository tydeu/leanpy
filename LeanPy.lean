/-
Copyright (c) 2023 Mac Malone. All rights reserved.
Released under the Apache 2.0 license as described in the file LICENSE.
syntax Authors Mac Malone
-/
import LeanPy.Data.Object.Basic
import LeanPy.Elab.BasicExprs
import LeanPy.Elab.Combinators
import LeanPy.Elab.Command
import LeanPy.Elab.Control
import LeanPy.Elab.Var
import LeanPy.ParserCategories
import LeanPy.Grammar
