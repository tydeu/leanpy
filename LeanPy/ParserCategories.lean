/-
Copyright (c) 2023 Mac Malone. All rights reserved.
Released under the Apache 2.0 license as described in the file LICENSE.
syntax Authors Mac Malone
-/
import LeanPy.Grammar
import LeanPy.Util.CompileParser

open Lean Parser

namespace LeanPy

open Internal

@[inline] private def mkParserCategories : ParserCategories := {}
@[inline] private def addParserCategory (categories : ParserCategories) (catName : Name) (category : ParserCategory) :=
  categories.insert catName category

private def parserPrios : NameMap Nat :=
  (∅ : NameMap _)
  |>.insert ``Grammar.assignment (eval_prio high)

local elab "compile_py_cats%" : command => do
  let cats := (parserExtension.getState (← getEnv)).categories
  let x := mkConst (``mkParserCategories)
  let x ← cats.foldlM (init := x) fun x catName cat => do
    let .str .anonymous catStr := catName
      | return x
    unless catStr.startsWith "py" do
      return x
    let declName ← compileParserCategory cat parserPrios `LeanPy.Grammar
    return mkApp3 (mkConst ``addParserCategory) x
      (toExpr catName) (mkConst declName)
  addAndCompileConstDef `LeanPy.parserCategories x ``ParserCategories

compile_py_cats%

local elab "mk_token_table%" : term =>
  let tks := mkTokenTableOfCategories parserCategories |>.values
  return tks.foldl (init := mkConst ``mkTokenTable) fun x tk =>
    mkApp2 (mkConst ``addNewToken) x (toExpr tk)

set_option compiler.extract_closed false in
def tokenTable : TokenTable := mk_token_table%
