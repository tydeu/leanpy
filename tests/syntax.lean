import LeanPy.Grammar

syntax withPosition("#print_py" LeanPy.Grammar.block) : command

macro_rules
| `(#print_py $blk) =>
  `(#eval IO.println $(Lean.quote <| toString blk))

#print_py
  x = f(1); 4 + f(3)
  if a is None:
    kill(pid)
    return True
  else:
    return False
