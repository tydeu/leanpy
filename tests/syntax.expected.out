(LeanPy.Grammar.block
 [(LeanPy.Grammar.stmt
   (LeanPy.Grammar.simpleStmts
    [(LeanPy.Grammar.assignment
      (group
       [(group (LeanPy.Grammar.starTargets [(LeanPy.Grammar.pyStarTarget__1 (LeanPy.Grammar.name `x))]) "=")]
       (LeanPy.Grammar.annotatedRhs
        (LeanPy.Grammar.starExprs
         [(LeanPy.Grammar.starExpr
           (LeanPy.Grammar.callExpr
            (LeanPy.Grammar.identExpr `f)
            "("
            [(LeanPy.Grammar.arguments
              (LeanPy.Grammar.args (group [(group (group (LeanPy.Grammar.numExpr (num "1"))))] [])))]
            ")"))]))
       []))
     ";"
     (LeanPy.Grammar.starExprs
      [(LeanPy.Grammar.starExpr
        (LeanPy.Grammar.add
         (LeanPy.Grammar.numExpr (num "4"))
         "+"
         (LeanPy.Grammar.callExpr
          (LeanPy.Grammar.identExpr `f)
          "("
          [(LeanPy.Grammar.arguments
            (LeanPy.Grammar.args (group [(group (group (LeanPy.Grammar.numExpr (num "3"))))] [])))]
          ")")))])]))
  (LeanPy.Grammar.stmt
   (LeanPy.Grammar.ifStmt
    "if"
    (LeanPy.Grammar.namedExpr
     (group (LeanPy.Grammar.is (LeanPy.Grammar.identExpr `a) "is" (LeanPy.Grammar.noneExpr "None"))))
    ":"
    (LeanPy.Grammar.block
     [(LeanPy.Grammar.stmt
       (LeanPy.Grammar.simpleStmts
        [(LeanPy.Grammar.starExprs
          [(LeanPy.Grammar.starExpr
            (LeanPy.Grammar.callExpr
             (LeanPy.Grammar.identExpr `kill)
             "("
             [(LeanPy.Grammar.arguments
               (LeanPy.Grammar.args (group [(group (group (LeanPy.Grammar.identExpr `pid)))] [])))]
             ")"))])]))
      (LeanPy.Grammar.stmt
       (LeanPy.Grammar.simpleStmts
        [(LeanPy.Grammar.returnStmt "return" [(LeanPy.Grammar.starExpr (LeanPy.Grammar.trueExpr "True"))])]))])
    []
    [(LeanPy.Grammar.elseBlock
      "else"
      ":"
      (LeanPy.Grammar.block
       [(LeanPy.Grammar.stmt
         (LeanPy.Grammar.simpleStmts
          [(LeanPy.Grammar.returnStmt
            "return"
            [(LeanPy.Grammar.starExpr (LeanPy.Grammar.falseExpr "False"))])]))]))]))])
